pkg/generator/plugin/modelgen/modelgen.go

```diff
@@ -0,0 +1,155 @@
+		if err != nil {
+			return err
+		}
+		constraints, err := getConstraintArgs(table)
+		if err != nil {
+			return err
+		}
+		modelType := snaker.SnakeToCamelIdentifier(tableName)
+		args.Tables = append(args.Tables, TableTemplateArgs{
+			TableName:              table.Table.TableName,
+			TableType:              snaker.ForceLowerCamelIdentifier(tableName),
+			TableSingletonName:     snaker.SnakeToCamelIdentifier(tableName),
+			ModelType:              modelType,
+			QualifiedModelType:     fmt.Sprintf("%s.%s", gen.modelPackage, modelType),
+			IsReferenceTable:       isReferenceTable(tableName),
+			ReferenceTableEnumType: getEnumTypeFromReferenceTableName(tableName),
+			Fields:                 fields,
+			Constraints:            constraints,
+		})
+	}
+	enumTemplate := utils.GetTemplate(gen.templateString)
+	return utils.RenderToFile(enumTemplate, gen.outputFile, args)
+}
+
+func getColumnToTypeMapping(
+	table metadata.Table,
+) map[string]string {
+	result := make(map[string]string)
+	for _, fk := range table.ForeignKeyConstraints {
+		if isReferenceTable(fk.ForeignTableName) {
+			result[fk.ColumnName] = getEnumTypeFromReferenceTableName(fk.ForeignTableName)
+		}
+	}
+	return result
+}
+
+func getFieldArgs(
+	data *metadata.Data, table metadata.Table,
+) ([]FieldTemplateArgs, error) {
+	columnToRefTableMapping := getColumnToTypeMapping(table)
+	var results []FieldTemplateArgs
+	for _, column := range table.Columns {
+		dataType, err := data.Loader.GetDataType(column.DataType)
+		if err != nil {
+			return nil, err
+		}
+		literal := dataType.Literal
+		if column.IsNullable {
+			literal = dataType.NullableLiteral
+		}
+		if enumName, ok := columnToRefTableMapping[column.ColumnName]; ok {
+			literal = enumName
+		} else if column.DataType == "USER-DEFINED" && column.UserDefinedTypeName != "citext" {
+			// citext is the only user-defined type that is not an enum
+			literal = snaker.SnakeToCamelIdentifier(column.UserDefinedTypeName)
+		}
+		results = append(results, FieldTemplateArgs{
+			Name:     column.ColumnName,
+			GooqType: dataType.Name,
+			Type:     literal,
+		})
+	}
+	return results, nil
+}
+
+func getConstraintArgs(
+	table metadata.Table,
+) ([]ConstraintTemplateArgs, error) {
+	var results []ConstraintTemplateArgs
+	for _, constraint := range table.Constraints {
+		var columns []string
+		err := json.Unmarshal([]byte(constraint.IndexKeys), &columns)
+		if err != nil {
+			return nil, err
+		}
+		for index := range columns {
+			column := columns[index]
+			columns[index] = strings.Replace(column, "\"", "\\\"", -1)
+		}
+		columnsString := fmt.Sprintf("{\"%s\"}", strings.Join(columns, "\",\""))
+		results = append(results, ConstraintTemplateArgs{
+			Name:      constraint.IndexName,
+			Columns:   columnsString,
+			Predicate: constraint.IndexPredicate,
+		})
+	}
+	return results, nil
+}
+
+func isReferenceTable(
+	tableName string,
+) bool {
+	return strings.HasSuffix(tableName, "_reference_table")
```

Comment:
We'll probably want to avoid hardcoding this suffix in the future.

I think reference tables are a great feature, but users might find it counter-intuitive; I don't think they would expect the table names to affect `gOOQ`'s codegen logic. We should be fine as long as we document this feature thoroughly.

Reply:
👍 will fix that

Reply:
I will add a task to document `reference_table` also